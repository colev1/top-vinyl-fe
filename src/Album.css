ul {
  list-style-type: none;
  padding-left: 5px;
}


.vinyl-container {
  background: #B4C5E4;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.track-container {
  text-align: left;
  display: flex;
}

.track-list {
  margin-top: 40px;
}

.track {
  border-bottom: 1px solid grey;
  margin-bottom: 15px;
  position: relative;
  height: 35px;
  display: flex;
  justify-content: space-between;
}

.track-name {
  display: inline-block;
  margin: 0 20px 0 0;
  max-height: 30px;
}

.delete-track-btn {
  position: relative;
  bottom: 5px;
}

.track-duration {
  display: inline-block;
  margin: 0 0 0 12px;
  font-weight: normal;
}

.album-info {
  margin-right: 30px;
}

.album-name {
  margin: 20px 0 0 0;
  padding: 0;
  font-size: 30px;
}

.album-genre {
  margin-top: 5px;
  color: #3F4756;
}

.artist {
  margin: 10px 0 0 0 ;
  color: #3F4756;
}

.view-albums-button {
  margin-bottom: 15px;
}
