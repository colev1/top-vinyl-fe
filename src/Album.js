import React, {Component} from 'react';

class Album extends Component {
  constructor(props) {
    super(props)
    this.state = {
      album: {},
      tracks: [],
      name: '',
      duration: ''
    }
  }

  componentDidMount() {
    if(this.props.albumId) {
      this.fetchTracks()
      this.fetchAlbum()
    }
  }

  fetchTracks = () => {
    const { albumId } = this.props; 
    fetch(`http://localhost:3000/api/v1/albums/${albumId}/tracks`)
      .then(response => response.json())
      .then(tracks => this.displayTracks(tracks))
      .catch(error => console.log(error))
  }

  fetchAlbum = () => {
    const { albumId } = this.props; 
    fetch(`http://localhost:3000/api/v1/albums/${albumId}`)
      .then(response => response.json())
      .then(album => {this.setState({album})})
      .catch(error => console.log(error))
  }

  displayTracks = (tracks) => {
    this.setState({tracks})
  }

  deleteTrack = (id) => {
    fetch(`http://localhost:3000/api/v1/tracks/${id}`, {
      method: 'DELETE'
    })
    .then(response => response.json())
    .then(result => this.fetchTracks())
    .catch(error => console.log(error))
  }

  addNewTrack = (e, albumId) => {
    e.preventDefault()
    let trackRequest = {
      name: this.state.name,
      duration: this.state.duration
    }
    fetch(`http://localhost:3000/api/v1/albums/${albumId}/tracks`, {
      method: 'POST',
      body: JSON.stringify(trackRequest),
      headers: {
        "Content-Type": "application/json"
      }
    })
    .then(response => response.json())
    .then(result => this.fetchTracks())
    .catch(error => console.log(error))
  }

  handleChange = (e) => {
    const {name, value} = e.target;
    this.setState({
      [name]: value
    })
  }

  render() {
    const {album, genre, artist, id} = this.state.album;
    let tracks = this.state.tracks.map(track => {
      return (
        <li key={track.id} className='track'> 
          <h4 className='track-name'> {track.name} <span className='track-duration'> {track.duration} </span></h4>
          <button className='delete-track-btn' onClick={()=> this.deleteTrack(track.id)}> delete</button>
        </li>
      )
    })

    return (
      <div className='vinyl-container'>
        <section className='track-container'>
          <article className='album-info'>
            <h1 className='album-name'> {album} </h1>
            <p className='artist'> by: {artist} </p>
            <p className='album-genre'> genre: {genre} </p>
          </article>
          <article className='track-list'>
            <ul> {tracks} </ul>
          </article>
        </section>
        <form onSubmit={(e) => this.addNewTrack(e, id)}> add a track:
          <input placeholder='name' onChange={this.handleChange}
          name='name' 
          value={this.state.name} />
          <input placeholder='duration' 
          name='duration'
          onChange={this.handleChange} value={this.state.duration} />
          <button type='submit'> submit </button>
        </form>
        <button className='view-albums-button' onClick={() => this.props.toggleDisplay()}>View All Ablums</button>
      </div>
    )
  }
}

export default Album;