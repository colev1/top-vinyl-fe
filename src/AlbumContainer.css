.albums-container {
  display: flex;
  flex-wrap: wrap;
}

.album {
  width: 300px;
}

h1:hover {
  color: red;
  cursor: pointer;
}