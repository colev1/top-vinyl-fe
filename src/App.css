.App {
  text-align: center;
  font-family: 'Inconsolata', monospace;
}

img {
  height: 80px;
  margin-right: 20px;
}

.title {
  /* height: 100px; */
  font-size: 44px;
  display: flex;
  justify-content: center;
  align-items: center;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
