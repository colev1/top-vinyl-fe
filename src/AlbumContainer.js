import React, {Component} from 'react';
import Album from './Album'
import './AlbumContainer.css';

class AlbumContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      displayTracks: false,
      albumId: null
    }
  }

  displayAlbumInfo = (id) => {
    console.log(id)
    this.setState({
      displayTracks: true,
      albumId: id
    })
  }
  
  render () {
  let albumsDisplay = this.props.albums.map(album => {
    return <div key={album.id} 
    className='album' 
    onClick={() => this.displayAlbumInfo(album.id)}>
      <h1> {album.album} </h1>
      <h3> {album.artist} </h3>
      <p> {album.genre} </p>
    </div>
    }
  )

  if(this.state.displayTracks) {
   return (
      <Album albumId={this.state.albumId}/>
   ) 
  } else {
    return (
      <div className='albums-container'>
        {albumsDisplay}
      </div>
    )
    }
  }
}

export default AlbumContainer;