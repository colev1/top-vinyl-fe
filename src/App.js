import React, { Component } from 'react';
import AlbumContainer from './AlbumContainer'
import './App.css';

class App extends Component {
  constructor () {
    super()
    this.state = {
      albums: [],
      error: '',
      album: '',
      genre: '',
      rating: '',
      artist: '',
      year: ''
    }
  }

  componentDidMount = () => {
    this.fetchAlbums()
  }

  fetchAlbums = () => {
    fetch('http://localhost:3000/api/v1/albums')
      .then(response => response.json())
      .then(result => this.addAlbums(result))
      .catch(error => console.log(error))
  }

  addAlbums = (albums) => {
    this.setState({
      albums
    })
  }

  deleteAlbum = (id) => {
    fetch(`http://localhost:3000/api/v1/albums/${id}`, {
      method: 'DELETE'
    })
    .then(response => response.json())
    .then(result => this.fetchAlbums())
    .catch(error => console.log(error))
  }

  handleChange = (e) => {
    const {name, value} = e.target;
    this.setState({
      [name]: value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    const {artist, genre, year, album} = this.state;
    const rating = parseInt(this.state.rating)
    const requestBody = {artist, genre, year, rating, album};
    fetch('http://localhost:3000/api/v1/albums', {
      method: 'POST',
      body: JSON.stringify(requestBody),
      headers: {
        "Content-Type": "application/json"
      }
    })
    .then(response => response.json())
    .then(result => console.log(result))
    .then(this.fetchAlbums())
    .catch(error => console.log(error))
  }

  render() {
    let {album, artist, year, genre, rating} = this.state;
    return (
      <div className="App">
        <h1> TOP VINYL </h1>

        <form onSubmit={(e) => this.handleSubmit(e)}> 
          add a new album:
          <input placeholder='album name' 
          name="album"
          value={album}
          onChange={this.handleChange} />
          <input placeholder='artist'
          name="artist"
          value={artist}
          onChange={this.handleChange} />
          <input placeholder='genre'
          name="genre"
          value={genre}
          onChange={this.handleChange} />
          <input placeholder='year'
          name="year"
          value={year}
          onChange={this.handleChange} />
          <input placeholder='rating'
          name="rating"
          value={rating}
          onChange={this.handleChange} />
          <button type="submit"
          > 
          add new album </button>
        </form>
        <AlbumContainer albums={this.state.albums} deleteAlbum={this.deleteAlbum}/>
      </div>
    );
  }
}

export default App;
