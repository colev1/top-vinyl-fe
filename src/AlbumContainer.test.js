import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import AlbumContainer from './AlbumContainer.js'
import Album from './Album.js'

describe('AlbumContainer', () => {
  let mockDelete
  let mockToggle
  let wrapper
  let mockAlbums

  beforeEach(() => {
    mockDelete = jest.fn()
    mockToggle = jest.fn()
    mockAlbums = [{
      album: 'This Album',
      artist: 'This Artist',
      year: '1991',
      rating: '3.33',
      genre: 'Pop'
    }]
    wrapper = shallow(<AlbumContainer
      albums={ mockAlbums }
      deleteAlbum={ mockDelete }
      toggleDisplay={ mockToggle }
      displayTracks={ false }
      />)
  })

  it('should match the snapshot', () => {
    
    expect(wrapper).toMatchSnapshot()
  })

  it('should render albums if props.displayTracks is false', () => {

    expect(wrapper.find('.albums-container').length).toBe(1)
  })

  it('should render an Album component if the displayTracks is true', async () => {
    await wrapper.setProps({
      displayTracks: true
    })
    
    expect(wrapper.find(Album).length).toBe(1)
  })


  it('displayAlbumInfo should setState and should call toggleDisplay', () => {
    wrapper.instance().displayAlbumInfo(5)

    expect(wrapper.state().albumId).toEqual(5)
    expect(mockToggle).toHaveBeenCalled()
  })
})