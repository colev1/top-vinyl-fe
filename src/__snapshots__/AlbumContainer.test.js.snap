// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumContainer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AlbumContainer
    albums={
      Array [
        Object {
          "album": "This Album",
          "artist": "This Artist",
          "genre": "Pop",
          "rating": "3.33",
          "year": "1991",
        },
      ]
    }
    deleteAlbum={[MockFunction]}
    displayTracks={false}
    toggleDisplay={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="album"
        >
          <h1
            onClick={[Function]}
          >
             
            THIS ALBUM
          </h1>
          <h3>
             
            This Artist
             
          </h3>
          <p>
             
            Pop
             
          </p>
          <button
            onClick={[Function]}
          >
             delete 
          </button>
        </div>,
      ],
      "className": "albums-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              onClick={[Function]}
            >
               
              THIS ALBUM
            </h1>,
            <h3>
               
              This Artist
               
            </h3>,
            <p>
               
              Pop
               
            </p>,
            <button
              onClick={[Function]}
            >
               delete 
            </button>,
          ],
          "className": "album",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "THIS ALBUM",
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "THIS ALBUM",
            ],
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "This Artist",
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "This Artist",
              " ",
            ],
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "Pop",
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "Pop",
              " ",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " delete ",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": " delete ",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="album"
          >
            <h1
              onClick={[Function]}
            >
               
              THIS ALBUM
            </h1>
            <h3>
               
              This Artist
               
            </h3>
            <p>
               
              Pop
               
            </p>
            <button
              onClick={[Function]}
            >
               delete 
            </button>
          </div>,
        ],
        "className": "albums-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                onClick={[Function]}
              >
                 
                THIS ALBUM
              </h1>,
              <h3>
                 
                This Artist
                 
              </h3>,
              <p>
                 
                Pop
                 
              </p>,
              <button
                onClick={[Function]}
              >
                 delete 
              </button>,
            ],
            "className": "album",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "THIS ALBUM",
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "THIS ALBUM",
              ],
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "This Artist",
                  " ",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "This Artist",
                " ",
              ],
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "Pop",
                  " ",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "Pop",
                " ",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " delete ",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": " delete ",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
